### PR TITLE
TRD tracking adaptions in view of GPU processing

### PR DESCRIPTION
**Pull request by martenole**
- tracklets are already sorted by HCID
- tracklets don't need to be loaded individually anymore

### CODE REVIEW

File: Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx [hunk 1]
```diff
@@ -197,25 +197,18 @@ void TRDGlobalTracking::run(ProcessingContext& pc)
...
   }
 
-  mTracker->SetTriggerRecordTimes(&(trdTriggerTimes[0]));
-  mTracker->SetTriggerRecordIndices(&(trdTriggerIndices[0]));
-  mTracker->SetNCollisions(nCollisions);
   //mTracker->DumpTracks();
   mTracker->ResetImpactAngleHistograms();
   mTracker->DoTracking(mChainTracking);
-  //mTracker->DumpTracks();
+  mTracker->DumpTracks();
```
Comment (davidrohr) [inline Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx]: leftover debug code, should not be part of the commit

File: GPU/GPUTracking/Global/GPUChainTrackingTRD.cxx [hunk 2]
```diff
@@ -38,11 +38,10 @@ int GPUChainTracking::RunTRDTracking()
   mRec->PushNonPersistentMemory(qStr2Tag("TRDTRACK"));
   SetupGPUProcessor(&Tracker, true);
 
-  for (unsigned int iTracklet = 0; iTracklet < mIOPtrs.nTRDTracklets; ++iTracklet) {
-    if (Tracker.LoadTracklet(mIOPtrs.trdTracklets[iTracklet], mIOPtrs.trdTrackletsMC ? mIOPtrs.trdTrackletsMC[iTracklet].mLabel : nullptr)) {
-      return 1;
-    }
+  if (mIOPtrs.trdTracklets == 0) {
```
Comment (davidrohr) [inline GPU/GPUTracking/Global/GPUChainTrackingTRD.cxx]: I don't understand this check. I would check if the number of tracklets = 0, we can return.
If the tracklet pointer is zero and the number is not, then it is a bug.

File: GPU/GPUTracking/Global/GPUChainTrackingTRD.cxx [hunk 3]
```diff
@@ -38,11 +38,10 @@ int GPUChainTracking::RunTRDTracking()
...
+    return 0;
   }
+  // TODO what about trdTrackletsMC? Are they needed at all?
```
Comment (davidrohr) [inline GPU/GPUTracking/Global/GPUChainTrackingTRD.cxx]: That already exists in the IOPtr, so you can just fill it from external as you do for the tracklet itself

Comment (martenole) [reply GPU/GPUTracking/Global/GPUChainTrackingTRD.cxx]: Yes, but that was originally added for AliRoot, but in the end I never used it. For o2 the labels are in a separate container anyways. So I think I will remove that entirely

File: Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx (hunk 1 above)
Comment (martenole) [reply Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx]: Done

File: GPU/GPUTracking/Global/GPUChainTrackingTRD.cxx (hunk 3 above)
Comment (davidrohr) [reply GPU/GPUTracking/Global/GPUChainTrackingTRD.cxx]: ok, makes sense

**Comment by martenole** [comment]
Debug output will go when this PR is not a draft anymore.
Next I need to make some adaptions for AliRoot. For O2 the code is in principle ready. The conversion of tracklets to space points is done externally by the `TrackletTransformer` and the space points are added to the ioPtrs

**Comment by davidrohr** [comment]
Hi @martenole : I have opened #6031, which adds GPU support for several more data types, in particular the TrackTPCITS, and of your GPUTRDTrack and GPUTRDTracker derived from O2 TrackParCov.
I checked that your PR rebases quite cleanly, there were only 3 small trivial collisions.

If nothing unexpected happens, my part should be done. It is compiling on the GPU now, and I can anyway only continue once this PR here is merged. So if you want you can rebase any time.

Then I got some new compiler warnings in your code, which you might want to fix:
```
/home/qon/alice/O2/GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx: In instantiation of ‘void o2::gpu::GPUTRDTracker_t<T, P>::DoTracking(o2::gpu::GPUChainTracking*) [with TRDTRK = o2::gpu::GPUTRDTrack_t<o2::gpu::trackInterface<o2::gpu::GPUTRDO2BaseTrack> >; PROP = o2::gpu::propagatorInterface<o2::base::PropagatorImpl<float> >]’:
/home/qon/alice/O2/GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:1503:16:   required from here
/home/qon/alice/O2/GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:208:29: warning: comparison of integer expressions of different signedness: ‘int’ and ‘const unsigned int’ [-Wsign-compare]
  208 |   for (int iColl = 0; iColl < GetConstantMem()->ioPtrs.nTRDTriggerRecords; ++iColl) {
      |                       ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/qon/alice/O2/GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:213:24: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]
  213 |       nTrklts = (iColl < GetConstantMem()->ioPtrs.nTRDTriggerRecords - 1) ? GetConstantMem()->ioPtrs.trdTrackletIdxFirst[iColl + 1] - GetConstantMem()->ioPtrs.trdTrackletIdxFirst[iColl] : GetConstantMem()->ioPtrs.nTRDTracklets - GetConstantMem()->ioPtrs.trdTrackletIdxFirst[iColl];
      |                 ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/qon/alice/O2/GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx: In instantiation of ‘int o2::gpu::GPUTRDTracker_t<T, P>::GetCollisionIDs(TRDTRK&, int*) const [with TRDTRK = o2::gpu::GPUTRDTrack_t<o2::gpu::trackInterface<o2::gpu::GPUTRDO2BaseTrack> >; PROP = o2::gpu::propagatorInterface<o2::base::PropagatorImpl<float> >]’:
/home/qon/alice/O2/GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:1503:16:   required from here
/home/qon/alice/O2/GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:524:29: warning: comparison of integer expressions of different signedness: ‘int’ and ‘const unsigned int’ [-Wsign-compare]
  524 |   for (int iColl = 0; iColl < GetConstantMem()->ioPtrs.nTRDTriggerRecords; ++iColl) {
```

**Comment by martenole** [comment]
In principle done. Waiting for https://github.com/AliceO2Group/AliceO2/pull/6031 and then will do a rebase on top.

File: GPU/GPUTracking/TRDTracking/GPUTRDTrackerComponent.cxx [hunk 4]
```diff
@@ -359,22 +360,14 @@ int GPUTRDTrackerComponent::DoEvent(const AliHLTComponentEventData& evtData, con
   fTracker->Reset();
   fChain->mIOPtrs.nMergedTracks = tracksTPC.size();
   fChain->mIOPtrs.nTRDTracklets = nTrackletsTotal;
-  fChain->AllocateIOMemory();
+  fChain->mIOPtrs.nTRDTriggerRecords = 1;
```
Comment (davidrohr) [inline GPU/GPUTracking/TRDTracking/GPUTRDTrackerComponent.cxx:370]: I think you also have to set `trdTriggerTimes` and `trdTrackletIdxFirst` to some valid values:
```
float dummyTime = 0;
int dummyIndex = 0;
fChain->mIOPtrs.trdTriggerTimes = &dummyTime;
fChain->mIOPtrs.trdTrackletIdxFirst = &dummyIndex;
```
or do I miss something? This will otherwise produce an array out of bounds.

**Review by davidrohr — CHANGES_REQUESTED** [review]

**Comment by davidrohr** [comment]
I have tried to look at run2 data after rebasing this on my branch.
My AliRoot reconstruction with TRD in the HLT crashes.
In my standalone benchmark, I see that except for sector 1 (where the TRD tracklets look OK), in all other sectors all TRD tracklets are seen in the outermost TRD layer. I think something in the conversion to space points goes wrong.

**Comment by davidrohr** [comment]
@martenole : Can you also move the RecoInputContainer to TRD/DataFormats?
Please don't add any additional dependencies to dataformats.
This is a header-only helper, so its user will have to pull in the required dependencies anyway. I just want to avoid to link to TRDReconstruction just to fetch this data.

**Comment by davidrohr** [comment]
> My AliRoot reconstruction with TRD in the HLT crashes.

OK, I am overly stupid... I had noted above that it will segfault, but I didn't apply my own patch... Anyhow, 2 more minor fixes were needed to make it run in AliRoot. Here is the full patch: 
[trd.patch.gz](https://github.com/AliceO2Group/AliceO2/files/6409423/trd.patch.gz)

Then I saw that for several setters of booleans, there is no way to unset them:
```
  GPUd() void SetGenerateSpacePoints() { mGenerateSpacePoints = true; }
  GPUd() void SetProcessPerTimeFrame() { mProcessPerTimeFrame = true; }
  GPUd() void EnableDebugOutput() { mDebugOutput = true; }
```
I would propose to change that to
```
  GPUd() void SetGenerateSpacePoints(bool v = true) { mGenerateSpacePoints = v; }
```

**Comment by davidrohr** [comment]
> In my standalone benchmark, I see that except for sector 1 (where the TRD tracklets look OK), in all other sectors all TRD tracklets are seen in the outermost TRD layer. I think something in the conversion to space points goes wrong.

So I followed this up a bit: Now that the reconstruction in AliRoot now longer crashes, the output looks also OK to me.
I tried again with the standalone benchmark, and before this PR it works, yielding the same result as AliRoot, but only after this PR I have the problem described above, that all TRD tracklets in sector != 0 are in the outermost layer.

**Comment by davidrohr** [comment]
screenshot1
screenshot1
In the display before and afterwards. So one sector seems ok. In the others, it is not onyl that everything is moved in the outermost layer, but also into a certain z region. The TRD is also seperated in different chambers in Z I assume.

**Comment by davidrohr** [comment]
```
GPU/GPUTracking/TRDTracking/macros/run_trd_tracker.C:  chainTracking->AllocateIOMemory();
```
I think we can get rid of this as well?

File: GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx [hunk 5]
```diff
@@ -603,14 +582,11 @@ GPUd() bool GPUTRDTracker_t<TRDTRK, PROP>::ConvertTrkltToSpacePoint(const GPUTRD
...
   matrix->LocalToMaster(xTrkltDet, xTrkltSec);
-  sp.mR = xTrkltSec[0];
-  sp.mX[0] = xTrkltSec[1];
-  sp.mX[1] = xTrkltSec[2];
-  sp.mDy = trklt.GetdY();
-
-  int modId = geo.GetSector(det) * GPUTRDGeometry::kNstack + geo.GetStack(det); // global TRD stack number
-  unsigned short volId = geo.GetGeomManagerVolUID(det, modId);
-  sp.mVolumeId = volId;
+  sp.setX(xTrkltSec[0]);
```
Comment (davidrohr) [inline GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:603]: In principle, there is no strict need that the SpacePoint is a class.
In O2 it is also allowed that we have structs, particularly when all members are accessible directly anyway, and when there is no complicated functionality inside.
In that case, you would not need the setters / getters. But it is up to you.

File: GPU/GPUTracking/TRDTracking/GPUTRDTrackerComponent.cxx [hunk 6]
```diff
@@ -359,22 +360,14 @@ int GPUTRDTrackerComponent::DoEvent(const AliHLTComponentEventData& evtData, con
...
-      if (fTracker->LoadTracklet(tracklets[iTracklet])) {
-        return -EINVAL;
-      }
-    } else {
-      if (fTracker->LoadTracklet(tracklets[iTracklet], trackletsMC[iTracklet].mLabel)) {
-        return -EINVAL;
-      }
-    }
-  }
+  std::sort(tracklets, tracklets + nTrackletsTotal);
```
Comment (davidrohr) [inline GPU/GPUTracking/TRDTracking/GPUTRDTrackerComponent.cxx]: This is actually illegal. You are reinterpret_casting the const away in the ptr casts above, which you are not allowed to do:
```
    if (blocks[iBlock].fDataType == (kAliHLTDataTypeTrack | kAliHLTDataOriginITS) && fRequireITStrack) {
      itsData = (AliHLTTracksData*)blocks[iBlock].fPtr;
      fBenchmark.AddInput(blocks[iBlock].fSize);
    } else if (blocks[iBlock].fDataType == (AliHLTTPCDefinitions::TracksOuterDataType() | kAliHLTDataOriginTPC)) {
      tpcData = (AliHLTTracksData*)blocks[iBlock].fPtr;
      fBenchmark.AddInput(blocks[iBlock].fSize);
    } else if (blocks[iBlock].fDataType == (kAliHLTDataTypeTrackMC | kAliHLTDataOriginTPC)) {
      tpcDataMC = (AliHLTTrackMCData*)blocks[iBlock].fPtr;
      fBenchmark.AddInput(blocks[iBlock].fSize);
    } else if (blocks[iBlock].fDataType == (AliHLTTRDDefinitions::fgkTRDTrackletDataType)) {
      tracklets = reinterpret_cast<GPUTRDTrackletWord*>(blocks[iBlock].fPtr);
      nTrackletsTotal = blocks[iBlock].fSize / sizeof(GPUTRDTrackletWord);
      fBenchmark.AddInput(blocks[iBlock].fSize);
    } else if (blocks[iBlock].fDataType == (AliHLTTRDDefinitions::fgkTRDMCTrackletDataType)) {
      hasMCtracklets = true;
      trackletsMC = reinterpret_cast<GPUTRDTrackletLabels*>(blocks[iBlock].fPtr);
      nTrackletsTotalMC = blocks[iBlock].fSize / sizeof(GPUTRDTrackletLabels);
      fBenchmark.AddInput(blocks[iBlock].fSize);
```
All these should be `reinterpret_cast<const ...>`.
And then this sort here is not possible.
You will have to create a local copy of the tracklets inside the tracker component, and then you can sort tha tracklets inside the copy. But you must not alter the input buffer.

**Comment by davidrohr** [comment]
For reference, the sorting was missing in the dump of Run 2 data, therefore processing of run2 data in the standalone benchmark was broken. I have fixed this in the standalone tool. No additional changes are needed in this PR besides those I already mentioned above.

File: Detectors/TRD/reconstruction/include/TRDReconstruction/RecoInputContainer.h [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+  unsigned int mNSpacePoints;
+  unsigned int mNTriggerRecords;
+  std::vector<float> trdTriggerTimes;
+  std::vector<int> trdTriggerIndices;
+};
+
+inline auto getRecoInputContainer(o2::framework::ProcessingContext& pc, o2::gpu::GPUTrackingInOutPointers& ptrs, const o2::globaltracking::RecoContainer* inputTracks)
+{
+  auto retVal = std::make_unique<RecoInputContainer>();
+  retVal->mTracksTPCITS = inputTracks->getTPCITSTracks<o2::dataformats::TrackTPCITS>();
```
Comment (davidrohr) [inline Detectors/TRD/reconstruction/include/TRDReconstruction/RecoInputContainer.h:55]: I am not 100% sure whether this is allowed like this. Do we have to check which inputs we have configured, or can we just try to read them, and if the input is not there, it will return an empty span?
But even then, let's assume the input is there for whatever reason even though we have not configured it in the sources on the command line, then we should not pick them up here.

File: Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx [hunk 8]
```diff
@@ -249,16 +185,14 @@ void TRDGlobalTracking::endOfStream(EndOfStreamContext& ec)
        mTimer.CpuTime(), mTimer.RealTime(), mTimer.Counter() - 1);
 }
 
-DataProcessorSpec getTRDGlobalTrackingSpec(bool useMC, bool useTrkltTransf, GTrackID::mask_t src)
+DataProcessorSpec getTRDGlobalTrackingSpec(bool useMC, GTrackID::mask_t src)
 {
   std::vector<OutputSpec> outputs;
 
   dataRequestTRD.requestTracks(src, false);
```
Comment (davidrohr) [inline Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx]: One more point. I don't really like that dataRequestTRD is a global variable.
This can be done more elegantly using a shared ptr, as I have done e.g. in https://github.com/AliceO2Group/AliceO2/blob/dev/GPU/Workflow/src/O2GPUDPLDisplay.cxx#L111
(Ping @shahor02 FYI)

File: Detectors/TRD/reconstruction/include/TRDReconstruction/RecoInputContainer.h (hunk 7 above)
Comment (shahor02) [reply Detectors/TRD/reconstruction/include/TRDReconstruction/RecoInputContainer.h:55]: if the input was not configured, RecoContainer will return an empty span. And since the input is configured by the command line of particular workflow, there should be no "unwanted" sources. In case you want to limit sources for matching (e.g. use only ITS-TPC, while the TPC should be anyway present in the input for the refit reason), you can see how I did this for ITS-TPC: https://github.com/shahor02/AliceO2/blob/1db020e89aa28df4f6e54d1da497b03469b3c8df/Detectors/GlobalTrackingWorkflow/src/tpcits-match-workflow.cxx#L67-L87

BTW, I see you don't use the RecoContainer::createTracks method, which accounts for redundancy of (in your case) in TPC and TPC-ITS sources. Here how I load the seeds from TPC and TPC-TOF tracks in the ITS-TPC matching: https://github.com/shahor02/AliceO2/blob/1db020e89aa28df4f6e54d1da497b03469b3c8df/Detectors/GlobalTracking/src/MatchTPCITS.cxx#L600-L629

File: GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx (hunk 5 above)
Comment (martenole) [reply GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:603]: In O2 `GPUTRDSpacePoint` is the same as `o2::trd::CalibratedTracklet`, which is a class. For AliRoot I needed the same interface, therefore I created a class. Of course we could also change `CalibratedTracklet` to be a simple struct. But since the class does not create any overhead I think it is fine to leave it the way it is

File: GPU/GPUTracking/TRDTracking/GPUTRDTrackerComponent.cxx (hunk 4 above)
Comment (martenole) [reply GPU/GPUTracking/TRDTracking/GPUTRDTrackerComponent.cxx:370]: Actually setting the trigger records is not needed for AliRoot. Even if they are not set you should not see a segfault. I only access those arrays in the tracker if `mProcessPerTimeFrame` is set to `true` which is not the case in AliRoot.

File: GPU/GPUTracking/TRDTracking/GPUTRDTrackerComponent.cxx (hunk 6 above)
Comment (martenole) [reply GPU/GPUTracking/TRDTracking/GPUTRDTrackerComponent.cxx]: Hm, with `reinterpret_cast<const ...>` the compiler is complaining about an invalid conversion:
/home/oschmidt/alice/sw/SOURCES/AliRoot/0_O2/0/GPU/GPUTracking/TRDTracking/GPUTRDTrackerComponent.cxx:292:21: error: invalid conversion from ‘const AliGPU::gpu::GPUTRDTrackletLabels*’ to ‘AliGPU::gpu::GPUTRDTrackletLabels*’ [-fpermissive]
  292 |       trackletsMC = reinterpret_cast<const GPUTRDTrackletLabels*>(blocks[iBlock].fPtr);

But I do create a local copy of the tracklets now and work with them instead of the input data block directly.

Comment (davidrohr) [reply GPU/GPUTracking/TRDTracking/GPUTRDTrackerComponent.cxx]: > Hm, with `reinterpret_cast<const ...>` the compiler is complaining about an invalid conversion:
> /home/oschmidt/alice/sw/SOURCES/AliRoot/0_O2/0/GPU/GPUTracking/TRDTracking/GPUTRDTrackerComponent.cxx:292:21: error: invalid conversion from ‘const AliGPU::gpu::GPUTRDTrackletLabels*’ to ‘AliGPU::gpu::GPUTRDTrackletLabels*’ [-fpermissive]

This is because you have the local definitions without const in line 274:
```
  GPUTRDTrackletWord* tracklets = nullptr;
  GPUTRDTrackletLabels* trackletsMC = nullptr;
```
These should be const as well.

**Comment by martenole** [comment]
> screenshot1
> 
> screenshot1
> In the display before and afterwards. So one sector seems ok. In the others, it is not onyl that everything is moved in the outermost layer, but also into a certain z region. The TRD is also seperated in different chambers in Z I assume.

Thank you very much for these checks. I think this was because I set the `VolumeID` for each space point to a dummy value and therefore all tracklets ended up in the same chamber for each sector. This should be fixed now. Indeed the TRD is separated along Z in 5 stacks.

File: GPU/GPUTracking/TRDTracking/GPUTRDTrackerComponent.cxx (hunk 6 above)
Comment (martenole) [reply GPU/GPUTracking/TRDTracking/GPUTRDTrackerComponent.cxx]: Of course, yes. Thanks a lot, this is fixed now

**Comment by martenole** [comment]
> ```
> GPU/GPUTracking/TRDTracking/macros/run_trd_tracker.C:  chainTracking->AllocateIOMemory();
> ```
> 
> I think we can get rid of this as well?

This macro was broken and I fixed it now as well. The only thing missing is setting the reconstruction options for the TRD inside the macro. Is this possible?
I tried via `GPUTRDTracker::Param().rec.trdPenaltyChi2 = 1234`, but `Param()` returns a const value and therefore I cannot change it that way

**Comment by davidrohr** [comment]
> nd I fixed it now as well. The only thing missing is setting the reconstruction options for the TRD inside

You should set
```
GPUSettingsRec cfgRec;
cfgRec.trdPenaltyChi2 = ...
```
before the call to `rec->SetSettings...`

File: Detectors/TRD/reconstruction/include/TRDReconstruction/RecoInputContainer.h (hunk 7 above)
Comment (martenole) [reply Detectors/TRD/reconstruction/include/TRDReconstruction/RecoInputContainer.h:55]: @shahor02 concerning the input configuration: the allowed sources (TPC,ITS-TPC) are already defined in the workflow: https://github.com/AliceO2Group/AliceO2/blob/f99e1c99b1e9777a9d29ee8146b8c97e33476df4/Detectors/TRD/workflow/src/trd-tracking-workflow.cxx#L52

The `createTracks` method I want to implement with a followup PR and get this merged first. Only a few minor things are missing now and it should be ready later today.

File: Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx (hunk 8 above)
Comment (martenole) [reply Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx]: Done

**Comment by martenole** [comment]
Ready once the CI is through. Next step is the implementation of the RecoContainer logic for the tracker inputs
